packages/visx-text/src/hooks/useText.ts

```diff
@@ -0,0 +1,122 @@
+import { useEffect, useState, useRef, useCallback } from 'react';
+import reduceCSSCalc from 'reduce-css-calc';
+import { usePreviousValue } from '@huse/previous-value';
+import { TextProps, WordsWithWidth, WordWithWidth } from '../types';
+import getStringWidth from '../util/getStringWidth';
+
+function isNumber(val: unknown): val is number {
+  return typeof val === 'number';
+}
+
+export function useText(props: TextProps): [WordsWithWidth[], string, string] {
+  const [wordsByLines, setWordsByLines] = useState<WordsWithWidth[]>([]);
+  const wordsWithWidth = useRef<WordWithWidth[]>([]);
+  const spaceWidth = useRef(0);
+  const prevProps = usePreviousValue(props);
+  const prevWordsByLines = usePreviousValue(wordsByLines);
+
+  const calculateWordsByLines = useCallback(
+    (lineWidth?: number) => {
+      const { scaleToFit } = props;
+      return wordsWithWidth.current.reduce((result: WordsWithWidth[], { word, width }) => {
+        const currentLine = result[result.length - 1];
+
+        if (
+          currentLine &&
+          (lineWidth == null ||
+            scaleToFit ||
+            (currentLine.width || 0) + width + spaceWidth.current < lineWidth)
+        ) {
+          // Word can be added to an existing line
+          currentLine.words.push(word);
+          currentLine.width = currentLine.width || 0;
+          currentLine.width += width + spaceWidth.current;
+        } else {
+          // Add first word to line or word is too long to scaleToFit on existing line
+          const newLine = { words: [word], width };
+          result.push(newLine);
+        }
+
+        return result;
+      }, []);
+    },
+    [props],
+  );
+
+  const updateWordsWithoutCalculate = useCallback(() => {
+    const words =
+      props.children == null ? [] : props.children.toString().split(/(?:(?!\u00A0+)\s+)/);
+    setWordsByLines([{ words }]);
+  }, [props]);
+
+  const updateWordsByLines = useCallback(
+    (needCalculate: boolean = false) => {
+      const { width, scaleToFit, children, style } = props;
+      // Only perform calculations if using features that require them (multiline, scaleToFit)
+      if (width || scaleToFit) {
+        if (needCalculate) {
+          const words: string[] =
+            children == null ? [] : children.toString().split(/(?:(?!\u00A0+)\s+)/);
+          wordsWithWidth.current = words.map(word => ({
+            word,
+            width: getStringWidth(word, style) || 0,
+          }));
+          spaceWidth.current = getStringWidth('\u00A0', style) || 0;
+        }
+
+        setWordsByLines(calculateWordsByLines(width));
+      } else {
+        updateWordsWithoutCalculate();
+      }
+    },
+    [calculateWordsByLines, props, updateWordsWithoutCalculate],
+  );
+
+  useEffect(() => {
```

Comment:
I think previously this check was needed because that's how it needed to be done inside a component. Hooks implement this dependency check for us, so I'm wondering if we can refactor this to compute `wordsByLines` behind a `useMemo` hook, with a dependency on `[children, style]` because that's when it should be updated. 

That would make this code more readable/easy to follow now that we no longer need this imperative logic.

Reply:
@williaster Hi Chris, please take a look, your idea is great. :+1: